File: game/lib/context/command.cpp
#include <context.h>
#include <fb/redis.h>
using namespace fb::game;

async::task<bool> fb::game::context::handle_command_map(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name  = parameters[0].asString();
    auto model = this->model.map.name2map(name);
    if (model == nullptr)
        co_return false;

    auto x = 0;
    auto y = 0;
    if (parameters.size() == 3)
    {
        if (parameters[1].isNumeric())
            x = parameters[1].asInt();

        if (parameters[2].isNumeric())
            y = parameters[2].asInt();
    }

    auto& map = this->maps[model->id];
    co_await session.map(&map, point16_t(x, y));
    co_return true;
}

async::task<bool> fb::game::context::handle_command_sound(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto value = parameters[0].asInt();
    this->send(session, fb::protocol::game::response::object::sound(session, SOUND(value)), scope::PIVOT);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_action(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto value = parameters[0].asInt();
    this->send(session, fb::protocol::game::response::session::action(session, ACTION(value), DURATION::SPELL), scope::PIVOT);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_weather(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto value = parameters[0].asInt();
    this->send(session, fb::protocol::game::response::weather(WEATHER_TYPE(value)), scope::PIVOT);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_bright(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto value = parameters[0].asInt();
    this->send(session, fb::protocol::game::response::bright(value), scope::PIVOT);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_timer(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto value = parameters[0].asInt();
    this->send(session, fb::protocol::game::response::timer(value), scope::PIVOT);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_effect(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto value = parameters[0].asInt();
    this->send(session, fb::protocol::game::response::object::effect(session, value), scope::PIVOT);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_disguise(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name = parameters[0].asString();
    auto mob  = this->model.mob.name2mob(name);
    if (mob == nullptr)
        co_return true;

    session.disguise(mob->look);
    this->send(session, fb::protocol::game::response::object::effect(session, 0x03), scope::PIVOT);
    this->send(session, fb::protocol::game::response::session::action(session, ACTION::CAST_SPELL, DURATION::SPELL), scope::PIVOT);
    this->send(session, fb::protocol::game::response::object::sound(session, SOUND::DISGUISE), scope::PIVOT);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_undisguise(fb::game::character& session, Json::Value& parameters)
{
    session.undisguise();
    co_return true;
}

async::task<bool> fb::game::context::handle_command_mob(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name  = parameters[0].asString();
    auto model = this->model.mob.name2mob(name);
    if (model == nullptr)
        co_return true;

    auto mob = model->make<fb::game::mob>(*this, fb::game::mob::config{.alive = true});
    auto map = session.map();
    co_await mob->map(map, session.position());
    co_return true;
}

async::task<bool> fb::game::context::handle_command_class(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name       = parameters[0].asString();
    auto class_type = CLASS::NONE;
    auto promotion  = uint8_t(0);
    if (this->model.promotion.name2class(name, class_type, promotion) == false)
        co_return true;

    session.cls(class_type);
    session.promotion(promotion);
    this->send(session, fb::protocol::game::response::session::id(session), scope::SELF);
    this->send(session, fb::protocol::game::response::session::state(session, STATE_LEVEL::LEVEL_MAX), scope::SELF);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_level(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto level = parameters[0].asInt();
    session.level(level);
    this->send(session, fb::protocol::game::response::session::state(session, STATE_LEVEL::LEVEL_MAX), scope::SELF);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_spell(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name  = parameters[0].asString();
    auto model = this->model.spell.name2spell(name);
    if (model == nullptr)
        co_return false;

    auto slot = co_await session.spells.add(*model);
    if (slot == 0xFF)
        co_return false;

    co_return true;
}

async::task<bool> fb::game::context::handle_command_item(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name  = parameters[0].asString();
    auto model = this->model.item.name2item(name);
    if (model == nullptr)
        co_return false;

    auto count = parameters.size() > 1 && parameters[1].isInt() ? parameters[1].asInt() : 1;

    auto item  = model->make(*this, count);
    co_await item->map(session.map(), session.position());
    co_return true;
}

async::task<bool> fb::game::context::handle_command_world(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name = parameters[0].asString();
    for (auto& [id, world] : this->model.world)
    {
        for (auto& [index, point] : world)
        {
            if (point.name == name)
            {
                session.send(fb::protocol::game::response::map::worlds(this->model, id, index));
                co_return true;
            }
        }
    }

    co_return false;
}

async::task<bool> fb::game::context::handle_command_script(fb::game::character& session, Json::Value& parameters)
{
    session.dialog.from("scripts/script.lua").func("func").pushobject(session).resume(1);

    co_return true;
}

async::task<bool> fb::game::context::handle_command_hair(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isInt() == false)
        co_return false;

    auto hair = parameters[0].asInt();
    session.look(hair);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_hair_color(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isInt() == false)
        co_return false;

    auto color = parameters[0].asInt();
    session.color(color);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_armor_color(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() == 0)
    {
        session.armor_color(std::nullopt);
        co_return true;
    }
    else if (parameters[0].isInt() == false)
    {
        co_return false;
    }
    else
    {
        auto color = parameters[0].asInt();
        session.armor_color(color);
        co_return true;
    }
}

async::task<bool> fb::game::context::handle_command_exit(fb::game::character& session, Json::Value& parameters)
{
    // this->_internal->send(fb::protocol::internal::request::shutdown());
    co_return true;
}

async::task<bool> fb::game::context::handle_command_tile(fb::game::character& session, Json::Value& parameters)
{
    auto listener = session.get_listener<fb::game::character>();
    auto map      = session.map();
    if (map == nullptr)
        co_return false;

    auto tile = (*map)(session.x(), session.y());
    if (tile == nullptr)
        co_return false;

    auto sstream = std::stringstream();
    sstream << "맵타일 : " << tile->id;
    if (listener != nullptr)
        listener->on_notify(session, sstream.str());

    sstream.str("");
    sstream << "오브젝트 : " << tile->object;
    if (listener != nullptr)
        listener->on_notify(session, sstream.str());

    co_return true;
}

async::task<bool> fb::game::context::handle_command_save(fb::game::character& session, Json::Value& parameters)
{
    co_await this->save(session);
    co_return true;
}

async::task<bool> fb::game::context::handle_command_mapobj(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto value = parameters[0].asInt();

    auto map   = session.map();
    if (map == nullptr)
        co_return false;

    (*map)(session.x(), session.y())->object = value;
    co_return true;
}

async::task<bool> fb::game::context::handle_command_randmap(fb::game::character& session, Json::Value& parameters)
{
    static std::vector<fb::game::map*> maps;
    static std::once_flag              flag;
    std::call_once(flag, [this] {
        std::srand(static_cast<uint32_t>(std::time(nullptr)));
        for (auto& [id, map] : this->maps)
        {
            maps.push_back(&map);
        }
    });

    auto  index = std::rand() % maps.size();
    auto& map   = maps[index];
    auto  x     = map->width() > 0 ? std::rand() % map->width() : 0;
    auto  y     = map->height() > 0 ? std::rand() % map->height() : 0;

    co_await session.map(map, point16_t(x, y));
    co_return true;
}

async::task<bool> fb::game::context::handle_command_npc(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isString() == false)
        co_return false;

    auto name  = parameters[0].asString();

    auto model = this->model.npc.name2npc(name);
    if (model == nullptr)
        co_return false;

    auto npc = model->make<fb::game::npc>(*this);
    npc->direction(session.direction());
    co_await npc->map(session.map(), session.position());
    co_return true;
}

async::task<bool> fb::game::context::handle_command_durability(fb::game::character& session, Json::Value& parameters)
{
    if (parameters.size() < 1)
        co_return false;

    if (parameters[0].isNumeric() == false)
        co_return false;

    auto percent    = std::max(0, std::min(100, (int)parameters[0].asUInt()));
    auto equipments = std::vector<fb::game::equipment*>();
    for (int i = 0; i < CONTAINER_CAPACITY; i++)
    {
        auto item = session.items[i];
        if (item == nullptr)
            continue;

        auto& model = item->based<fb::model::item>();
        if (model.attr(ITEM_ATTRIBUTE::EQUIPMENT) == false)
            continue;

        auto equipment = static_cast<fb::game::equipment*>(item);
        equipments.push_back(equipment);
    }

    for (auto& [parts, item] : session.items.equipments())
    {
        if (item == nullptr)
            continue;

        auto equipment = static_cast<fb::game::equipment*>(item);
        equipments.push_back(equipment);
    }

    for (auto equipment : equipments)
    {
        auto& model = equipment->based<fb::model::equipment>();
        equipment->durability(uint16_t(model.durability * (percent / 100.0f)));
    }

    co_return true;
}

async::task<bool> fb::game::context::handle_command_concurrency(fb::game::character& session, Json::Value& parameters)
{
    auto seconds = parameters.size() >= 1 && parameters[0].isNumeric() ? parameters[0].asInt() : 10;
    auto key     = parameters.size() >= 2 && parameters[1].isString() ? parameters[1].asString() : "global";

    try
    {
        co_await this->_mutex.sync<bool>(key, [this, &session, seconds](auto& trans) -> async::task<bool> {
            for (int i = 0; i < seconds; i++)
            {
                session.chat(std::format("{}초 후에 풀립니다.", seconds - i));
                co_await this->sleep(1s);
            }
            co_return true;
        });
    }
    catch (fb::lock_error&)
    {
        session.chat("리소스 점유 실패");
    }
    catch (std::exception& e)
    {
        session.chat(e.what());
    }

    co_return true;
}

async::task<bool> fb::game::context::handle_command_sleep(fb::game::character& session, Json::Value& parameters)
{
    auto seconds = parameters.size() >= 1 && parameters[0].isNumeric() ? parameters[0].asInt() : 10;
    co_await this->sleep(std::chrono::seconds{seconds});
    co_return true;
}

File: game/scripts/command.lua
function string:split(delimiter)
    local result = { }
    local from  = 1
    local delim_from, delim_to = string.find( self, delimiter, from  )
    while delim_from do
        table.insert( result, string.sub( self, from , delim_from-1 ) )
        from  = delim_to + 1
        delim_from, delim_to = string.find( self, delimiter, from  )
    end
    table.insert( result, string.sub( self, from  ) )
    return result
end

command_funcs = {
    ['관리자'] =
    function (me, args)
        me:mkspell('강제이동(좌)')
        me:mkspell('강제이동(하)')
        me:mkspell('강제이동(우)')
        me:mkspell('강제이동(상)')
        return true
    end,

    ['경험치'] =
    function (me, args)
        local exp = table.unpack(args)
        exp = tonumber(exp)
        me:exp(exp)
        return true
    end,

    ['맵이동'] =
    function (me, args)
        local map, x, y = table.unpack(args)
        if name2map(map) == nil then
            return false
        end

        if x ~= nil and y ~= nil then
            me:map(map, tonumber(x), tonumber(y))
        else
            me:map(map)
        end
        return true
    end,

    ['사운드'] =
    function (me, args)
        local sound = table.unpack(args)
        me:sound(tonumber(sound))
        return true
    end,

    ['액션'] = 
    function (me, args)
        local action = table.unpack(args)
        me:action(tonumber(action))
        return true
    end,

    ['날씨'] = 
    function (me, args)
        local value = table.unpack(args)
        weather(tonumber(value))
        return true
    end,

    ['밝기'] = 
    function (me, args)
        local value = table.unpack(args)
        bright(tonumber(value))
        return true
    end,

    ['타이머'] = 
    function (me, args)
        local time = table.unpack(args)
        timer(tonumber(time), true)
        return true
    end,

    ['타이틀'] = 
    function (me, args)
        local value = table.unpack(args)
        me:title(tonumber(value))
        return true
    end,

    ['이펙트'] =
    function (me, args)
        local value = table.unpack(args)
        me:effect(tonumber(value))
        return true
    end,

    ['변신'] = 
    function (me, args)
        local value = table.unpack(args)
        me:disguise(tonumber(value))
        return true
    end,

    ['변신해제'] = 
    function (me, args)
        me:disguise(nil)
        return true
    end,

    ['마법배우기'] =
    function (me, args)
        local name = table.unpack(args)
        me:mkspell(name)
        return true
    end,

    ['마법지우기'] = 
    function (me, args)
        if #args == 0 then
            me:rmspell()
        else
            local slot = table.unpack(args)
            me:rmspell(slot)
        end
        return true
    end,

    ['몬스터생성'] = 
    function (me, args)
        local name = table.unpack(args)
        local x, y = me:position()
        me:spawn_mob(name, x, y, false)
        return true
    end,

    ['직업바꾸기'] = 
    function (me, args)
        local name = table.unpack(args)
        local class, promotion = name2class(name)
        if class ~= nil then
            me:class(class)
            me:promotion(promotion)
        end
        return true
    end,

    ['레벨바꾸기'] = 
    function (me, args)
        local level = table.unpack(args)
        me:level(tonumber(level))
        return true
    end,

    ['힘바꾸기'] = 
    function (me, args)
        local value = table.unpack(args)
        me:base_str(tonumber(value))
        return true
    end,

    ['민첩바꾸기'] = 
    function (me, args)
        local value = table.unpack(args)
        me:base_dex(tonumber(value))
        return true
    end,

    ['지력바꾸기'] = 
    function (me, args)
        local value = table.unpack(args)
        me:base_int(tonumber(value))
        return true
    end,

    ['체력바꾸기'] = 
    function (me, args)
        local value = table.unpack(args)
        me:base_hp(tonumber(value))
        me:hp(me:maxhp())
        if me:state() == STATE_GHOST then
            me:state(STATE_NORMAL)
        end
        return true
    end,

    ['마력바꾸기'] = 
    function (me, args)
        local value = table.unpack(args)
        me:base_mp(tonumber(value))
        me:mp(me:maxmp())
        return true
    end,

    ['아이템생성'] = 
    function (me, args)
        local name, count = table.unpack(args)
        me:mkitem(name, tonumber(count))
        return true
    end,

    ['월드맵'] = 
    function (me, args)
        local name = table.unpack(args)
        me:world(name)
        return true
    end,

    ['스크립트'] = 
    function (me, args)
        me:script("script.lua", "func", 1, "hello", "good")
        return true
    end,

    ['머리바꾸기'] = 
    function (me, args)
        local value = table.unpack(args)
        me:look(tonumber(value))
        return true
    end,

    ['머리염색'] = 
    function (me, args)
        local value = table.unpack(args)
        me:color(tonumber(value))
        return true
    end,

    ['갑옷염색'] = 
    function (me, args)
        local value = table.unpack(args)
        me:armor_color(tonumber(value))
        return true
    end,

    ['서버종료'] = 
    function (me, args)
        me:message('미구현')
        return true
    end,

    ['맵타일'] = 
    function (me, args)
        local map = me:map()
        if map == nil then
            return true
        end

        local x, y = me:position()
        if #args == 0 then
            local id, obj, blocked = map:tile(x, y)
            if id == nil then
                return true
            end

            if blocked then
                blocked = 'true'
            else
                blocked = 'false'
            end
            me:message(string.format('타일 : %d\n오브젝트 : %d\n블록 : %s', id, obj, blocked), MESSAGE_TYPE_POPUP)
        else
            local value = table.unpack(args)
            map:tile(x, y, tonumber(value))
        end
        return true
    end,

    ['서버저장'] = 
    function (me, args)
        save()
        return true
    end,

    ['랜덤이동'] = 
    function (me, args)
        local maps = maps()
        local map = maps[math.random(0, #maps-1)]
        local x = math.random(0, map:width())
        local y = math.random(0, map:height())
        me:map(map, x, y)
        return true
    end,

    ['엔피씨생성'] = 
    function (me, args)
        local name, map, x, y = table.unpack(args)
        if map == nil then
            map = me:map()
        else
            map = name2map(name)
        end

        if x == nil or y == nil then
            x, y = me:position()
        else
            x = tonumber(x)
            y = tonumber(y)
        end
        mknpc(name, map, x, y)
        return true
    end,

    ['내구도'] = 
    function (me, args)
        local percent = table.unpack(args)
        percent = tonumber(percent)
        percent = math.max(0, math.min(100, tonumber(percent)))
        for parts, equipment in pairs(me:equipments()) do
            local model = equipment:model()
            equipment:durability(model:durability() * (percent / 100.0))
        end

        for _, item in pairs(me:items()) do
            local model = item:model()
            if model:attr(ITEM_ATTRIBUTE_EQUIPMENT) then
                item:durability(model:durability() * (percent / 100.0))
            end
        end
        return true
    end,

    ['sleep'] = 
    function (me, args)
        local time = table.unpack(args)
        sleep(tonumber(time))
        me:message('done')
        return true
    end,

    ['광고'] = 
    function (me, args)
        local width, height, url, time = table.unpack(args)
        me:ad(tonumber(width), tonumber(height), url, tonumber(time))
        return true
    end,

    ['웹'] = 
    function (me, args)
        local type, url, message = table.unpack(args)
        me:web(tonumber(type), url)
        return true
    end,

    ['메일쓰기'] = 
    function (me, args)
        local to, title, contents = table.unpack(args)
        me:send_mail(to, title, contents)
        return true
    end,

    ['쿨타임초기화'] = 
    function (me, args)
        for slot, spell in pairs(me:spells()) do
            me:spell(spell):delay(0)
        end
        return true
    end,

    ['금전'] = 
    function (me, args)
        local money = table.unpack(args)
        me:money(me:money() + tonumber(money))
        return true
    end,

    ['성전환'] = 
    function (me, args)
        if me:sex() == SEX_MAN then
            me:sex(SEX_WOMAN)
        else
            me:sex(SEX_MAN)
        end
        return true
    end,

    ['서버종료'] = 
    function (me, args)
        local delay = table.unpack(args)
        if delay == nil then
            delay = 10
        else
            delay = tonumber(delay)
        end

        while delay > 0 do
            if delay > 600 then
                term = math.min(600, delay-600)
            elseif delay > 60 then
                term = math.min(60, delay-60)
            elseif delay > 10 then
                term = math.min(10, delay-10)
            else
                term = 1
            end

            times = delay
            hours = times // 3600
            times = times - (hours * 3600)

            mins = times // 60
            times = times - (mins * 60)

            secs = times % 60
            message = ''
            if hours > 0 then
                message = string.format('%d시간', hours)
            end

            if mins > 0 then
                message = string.format('%s %d분', message, mins)
            end

            if secs > 0 then
                message = string.format('%s %d초', message, secs)
            end

            broadcast(string.format('[공지] %s 후 서버가 종료됩니다.', message), MESSAGE_TYPE_NOTIFY, BROADCAST_TYPE_GLOBAL)
            sleep(term * 1000)
            delay = delay - term
        end
        shutdown()
        return true
    end
}